src/core/public/chrome/utils.ts

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { AppCategory } from 'opensearch-dashboards/public';
+import { groupBy, sortBy } from 'lodash';
+import { ChromeNavLink } from './nav_links';
+import { ChromeRegistrationNavLink } from './nav_group';
+
+export const LinkItemType = {
+  LINK: 'link',
+  CATEGORY: 'category',
+  PARENT_LINK: 'parentLink',
+} as const;
+
+export type LinkItem = { order?: number } & (
+  | { itemType: 'link'; link: ChromeNavLink & ChromeRegistrationNavLink }
+  | { itemType: 'parentLink'; link?: ChromeNavLink & ChromeRegistrationNavLink; links: LinkItem[] }
+  | { itemType: 'category'; category?: AppCategory; links?: LinkItem[] }
+);
+
+export function getAllCategories(allCategorizedLinks: Record<string, ChromeNavLink[]>) {
+  const allCategories = {} as Record<string, AppCategory | undefined>;
+
+  for (const [key, value] of Object.entries(allCategorizedLinks)) {
+    allCategories[key] = value[0].category;
+  }
+
+  return allCategories;
+}
+
+export function fulfillRegistrationLinksToChromeNavLinks(
+  registerNavLinks: ChromeRegistrationNavLink[],
+  navLinks: ChromeNavLink[]
+): Array<ChromeNavLink & ChromeRegistrationNavLink> {
+  const allExistingNavLinkId = navLinks.map((link) => link.id);
+  return (
+    registerNavLinks
+      ?.filter((navLink) => allExistingNavLinkId.includes(navLink.id))
+      .map((navLink) => ({
+        ...navLinks[allExistingNavLinkId.indexOf(navLink.id)],
+        ...navLink,
+      })) || []
+  );
+}
+
+export const getOrderedLinks = (navLinks: ChromeNavLink[]): ChromeNavLink[] =>
+  sortBy(navLinks, (link) => link.order);
+
+export function flattenLinksOrCategories(linkItems: LinkItem[]): ChromeNavLink[] {
+  return linkItems.reduce((acc, item) => {
+    if (item.itemType === LinkItemType.LINK) {
+      acc.push(item.link);
+    } else if (item.itemType === LinkItemType.PARENT_LINK) {
+      if (item.link) {
+        acc.push(item.link);
+      }
+      acc.push(...flattenLinksOrCategories(item.links));
+    } else if (item.itemType === LinkItemType.CATEGORY) {
+      acc.push(...flattenLinksOrCategories(item.links || []));
+    }
+
+    return acc;
+  }, [] as ChromeNavLink[]);
+}
+
+export const generateItemTypeByLink = (
+  navLink: ChromeNavLink & ChromeRegistrationNavLink,
+  navLinksGoupedByParentNavLink: Record<string, ChromeNavLink[]>
+): LinkItem => {
+  const navLinksUnderParentId = navLinksGoupedByParentNavLink[navLink.id];
+
+  if (navLinksUnderParentId) {
+    return {
+      itemType: LinkItemType.PARENT_LINK,
+      link: navLink,
+      links: getOrderedLinks(navLinksUnderParentId || []).map((navLinkUnderParentId) =>
+        generateItemTypeByLink(navLinkUnderParentId, navLinksGoupedByParentNavLink)
+      ),
+      order: navLink?.order,
```

Comment:
Nit: this optional chain seems unnecessary

Reply:
Sure, remove that.